src/Processors/QueryPlan/Optimizations/optimizeUseProjections.cpp

```diff
@@ -0,0 +1,1192 @@
+
+    if (!has_nornal_parts)
+    {
+        /// All parts are taken from projection
+        aggregating->requestOnlyMergeForAggregateProjection(expr_or_filter_node.step->getOutputStream());
+        node.children.front() = &expr_or_filter_node;
+    }
+    else
+    {
+        node.step = aggregating->convertToAggregatingProjection(expr_or_filter_node.step->getOutputStream());
+        node.children.push_back(&expr_or_filter_node);
+    }
+
+    return true;
+}
+
+
+static ActionsDAGPtr makeMaterializingDAG(const Block & proj_header, const Block main_header)
+{
+    /// Materialize constants in case we don't have it in output header.
+    /// This may happen e.g. if we have PREWHERE.
+
+    size_t num_columns = main_header.columns();
+    /// This is a error; will have block structure mismatch later.
+    if (proj_header.columns() != num_columns)
+        return nullptr;
+
+    std::vector<size_t> const_positions;
+    for (size_t i = 0; i < num_columns; ++i)
+    {
+        auto col_proj = proj_header.getByPosition(i).column;
+        auto col_main = main_header.getByPosition(i).column;
+        bool is_proj_const = col_proj && isColumnConst(*col_proj);
+        bool is_main_proj = col_main && isColumnConst(*col_main);
+        if (is_proj_const && !is_main_proj)
+            const_positions.push_back(i);
+    }
+
+    if (const_positions.empty())
+        return nullptr;
+
+    ActionsDAGPtr dag = std::make_unique<ActionsDAG>();
+    auto & outputs = dag->getOutputs();
+    for (const auto & col : proj_header.getColumnsWithTypeAndName())
+        outputs.push_back(&dag->addInput(col));
+
+    for (auto pos : const_positions)
+    {
+        auto & output = outputs[pos];
+        output = &dag->materializeNode(*output);
+    }
+
+    return dag;
+}
+
+static bool hasAllRequiredColumns(const ProjectionDescription * projection, const Names & required_columns)
+{
+    for (const auto & col : required_columns)
+    {
+        if (!projection->sample_block.has(col))
+            return false;
+    }
+
+    return true;
+}
+
+bool optimizeUseNormalProjections(Stack & stack, QueryPlan::Nodes & nodes)
+{
+    const auto & frame = stack.back();
+
+    auto * reading = typeid_cast<ReadFromMergeTree *>(frame.node->step.get());
+    if (!reading)
+        return false;
+
+    if (!canUseProjectionForReadingStep(reading))
+        return false;
+
+    auto iter = stack.rbegin();
+    while (iter != stack.rend())
+    {
+        auto next = std::next(iter);
+        if (next == stack.rend())
+            break;
+
+        if (!typeid_cast<FilterStep *>(next->node->step.get()) &&
+            !typeid_cast<ExpressionStep *>(next->node->step.get()))
+            break;
+
+        iter = next;
+    }
+
+    if (iter == stack.rbegin())
+        return false;
+
+    const auto metadata = reading->getStorageMetadata();
+    const auto & projections = metadata->projections;
+
+    std::vector<const ProjectionDescription *> normal_projections;
```

Comment:
`ordinary_projections`

Reply:
Currently, we call them normal. And have a special `ProjectionDescription::Type::Normal`.
I won't change this naming in this pr